cross-project `ref` + model `access` + `dependencies.yml` (#3577)
dbt-labs/docs.getdbt.com: website/docs/docs/collaborate/govern/project-dependencies.md

---
title: "Project dependencies"
id: project-dependencies
sidebar_label: "Project dependencies"
description: "Reference public models across dbt projects"
---

:::info Cross-project `ref`
This is a feature of dbt Cloud Enterprise.
:::

> dbt Core v1.6 introduces a notion of `dependencies` between dbt projects. You're likely already familiar with installing other projects as [packages](/docs/build/packages), whereby you pull down another project's source code and treat it as your own.

Reviewer: ```suggestion
dbt Core v1.6 introduces a notion of `dependencies` between dbt projects. You might already be familiar with installing other projects as [packages](/docs/build/packages), whereby you pull down another project's source code and treat it as your own.
```

> There is a new kind of `project` dependency. Both dependencies can be defined in `dependencies.yml`:

Reviewer: It feels like there is gap between the opening sentences and this one. Is there some more information we want to share to help users connect the dots here?

Author: writing more here!

```yml
packages:
  - package: dbt-labs/dbt_utils
    version: 1.1.1

projects:

>   - project: jaffle_finance

Reviewer 2: This is actually `name:`, not `project:`

Author: I'm totally okay with `name`! will update it here

>   - project: jaffle_finance

Reviewer: ```suggestion
  - name: jaffle_finance
```

```

> The first is familiar: I want to use macros from the `dbt_utils` package, and so I pull it down as source code. I can then call any macro from the package, just as if it were a macro defined in my own project.

Reviewer: ```suggestion
The first is familiar: you want to use macros from the `dbt_utils` package, so you pull it down as source code. You can then call any macro from the package as if it were a macro defined in your project.
```

> The second is new. Unlike installing a package, the models in the `jaffle_finance` project will not be pulled down as source code, or selected to run during `dbt run`. Instead, dbt Cloud will provide `dbt-core` with expect stateful input that enables it to resolve references to public models in the `jaffle_finance` project—and _only_ the public models.

Reviewer: ```suggestion
The second is new. Unlike installing a package, the models in the `jaffle_finance` project will not be pulled down as source code or selected to run during `dbt run`. Instead, dbt Cloud will provide `dbt-core` with expected stateful input that enables it to resolve references to public models in the `jaffle_finance` project—and _only_ the public models.
```

> The second is new. Unlike installing a package, the models in the `jaffle_finance` project will not be pulled down as source code, or selected to run during `dbt run`. Instead, dbt Cloud will provide `dbt-core` with expect stateful input that enables it to resolve references to public models in the `jaffle_finance` project—and _only_ the public models.

Reviewer 2: 👀 "with expect stateful input" -> "with expected stateful input"

Models referenced from a `project`-type dependency must use [two-argument `ref`](/reference/dbt-jinja-functions/ref#two-argument-variant), including the project name. Only public models can be accessed in this way.

> It is equally possible to install the `jaffle_finance` project as a `package` dependency. This will pull down its full source code and require dbt to parse all its contents. dbt will expect you to configure and run those models as your own. This can be a useful pattern to achieve certain types of unified deployments in production or to make a coordinated change across multiple projects in development. However, it can add significant complexity and latency when working within the narrower scope of a single project.

Author: > However, it can add significant complexity and latency when working within the narrower scope of a single project.

and cost!

> It is equally possible to install the `jaffle_finance` project as a `package` dependency. This will pull down its full source code and require dbt to parse all its contents. dbt will expect you to configure and run those models as your own. This can be a useful pattern to achieve certain types of unified deployments in production or to make a coordinated change across multiple projects in development. However, it can add significant complexity and latency when working within the narrower scope of a single project.

Reviewer: ```suggestion
It is equally possible to install the `jaffle_finance` project as a `package` dependency. This will pull down its full source code and require dbt to parse all its contents. dbt will expect you to configure and run those models as your own. This can be a useful pattern to achieve certain types of unified deployments in production or to make a coordinated change across multiple projects in development. However, it can add significant complexity, latency, and cost when working within the narrower scope of a single project.
```

> It is equally possible to install the `jaffle_finance` project as a `package` dependency. This will pull down its full source code, and require dbt to parse all its contents. dbt will expect you to configure and run those models as your own. This can be a useful pattern to achieve certain types of unified deployments in production, or to make a coordinated change across multiple projects in development. It can add significant complexity and latency, however, when working within the narrower scope of a single project.

Reviewer: ```suggestion
It is equally possible to install the `jaffle_finance` project as a `package` dependency. This will pull down its full source code and require dbt to parse all its contents. dbt will expect you to configure and run those models as your own. This can be a useful pattern to achieve certain types of unified deployments in production or to make a coordinated change across multiple projects in development. However, it can add significant complexity and latency when working within the narrower scope of a single project.
```

---
title: "Project dependencies"
id: project-dependencies
sidebar_label: "Project dependencies"
description: "Reference public models across dbt projects"
---

> 

Author: TODO: Add a note to this page about project-level cycle detection (https://github.com/dbt-labs/dbt-core/issues/7468)

:::info

> "Project" dependencies and cross-project `ref` are features of dbt Cloud Enterprise, currently available in closed beta. Please reach out to your account team!

Reviewer 3: ```suggestion
"Project" dependencies and cross-project `ref` is currently in closed beta and are features of dbt Cloud Enterprise. To access these features, please contact your account team.
```

:::

For a long time, dbt has supported code reuse and extension by installing other projects as [packages](/docs/build/packages). When you install another project as a package, you are pulling in its full source code, and adding it to your own. This enables you to call macros and run models defined in that other project.

While this is a great way to reuse code, share utility macros, and establish a starting point for common transformations, it's not a great way to enable collaboration across teams and at scale, especially at larger organizations.

This year, dbt Labs is introducing an expanded notion of `dependencies` across multiple dbt projects:

> - **Packages** are the familiar & preexisting type of dependency. You take this dependency by installing the package's full source code (like a software library).

Reviewer 3: ```suggestion
- **Packages** &mdash; Familiar and pre-existing type of dependency. You take this dependency by installing the package's full source code (like a software library).
```

> - **Projects** are a new way to take a dependency on another project. Using a metadata service that runs behind the scenes, dbt Cloud resolves references on-the-fly to public models defined in other projects. Since you don't need to parse or run those upstream models yourself, your dependency on those models is as an API that returns a dataset. It's the responsibility of that public model's maintainer to guarantee its quality and stability.

Reviewer 3: ```suggestion
- **Projects** &mdash; A _new_ way to take a dependency on another project. Using a metadata service that runs behind the scenes, dbt Cloud resolves references on-the-fly to public models defined in other projects. You don't need to parse or run those upstream models yourself. Instead, you treat your dependency on those models as an API that returns a dataset. The maintainer of the public model is responsible for guaranteeing its quality and stability.
```

Author: yes!

> 

Reviewer 3: sugg adding a header here
```suggestion
## Use case example

```

> Say I work on the Marketing team at the Jaffle Shop. The name of my project is `jaffle_marketing`:

Reviewer 3: ```suggestion
As an example, let's say I work on the Marketing team at the Jaffle Shop. The name of my project is `jaffle_marketing`:
```

> As an example, let's say I work on the Marketing team at the Jaffle Shop. The name of my project is `jaffle_marketing`:

Reviewer 3: ```suggestion
As an example, let's say you work on the Marketing team at the Jaffle Shop. The name of the project is `jaffle_marketing`:
```

<File name="dbt_project.yml">

```yml
name: jaffle_marketing
```

</File>

> I want to take a dependency on two other projects:

Reviewer 3: ```suggestion
To take a dependency on two other projects:
```

> - `dbt_utils`: public & open source, maintained by dbt Labs, a collection of utility macros

Reviewer 3: ```suggestion
- `dbt_utils` as a [package](#packages-use-case): An open-source and public collection of utility macros. It's maintained by dbt Labs.
```

> - `jaffle_finance`: data models about the Jaffle Shop's revenue, maintained by our colleagues on the Finance team

Reviewer 3: ```suggestion
- `jaffle_finance` as a [project use-case](#projects-use-case): Data models about the Jaffle Shop's revenue. It's maintained by our colleagues on the Finance team
```

<File name="dependencies.yml">

```yml
packages:
  - package: dbt-labs/dbt_utils
    version: 1.1.1

projects:
  - name: jaffle_finance  # matches the 'name' in their 'dbt_project.yml'
```

</File>

> The first is familiar: I want to use some of the macros that are already defined in `dbt_utils`. I pull down its full contents (100+ macros) as source code, and add them to my environment. I can then call any macro from the package, just as I can call macros that I define in my own project.

Reviewer 3: ```suggestion
- The `dbt_utils` package &mdash; This is a familiar scenario, you want to use some of the macros that are already defined in `dbt_utils`. You pull down its full contents (100+ macros) as source code and add them to your environment. You can then call any macro from the package, just as you can call macros defined in your own project.
```

> The second is new. Unlike installing a package, the models in the `jaffle_finance` project will not be pulled down as source code, and parsed into my project. Instead, dbt Cloud provides a metadata service that resolves references to [**public models**](/docs/collaborate/govern/model-access) defined in the `jaffle_finance` project.

Reviewer 3: ```suggestion
-  The `jaffle_finance` projects &mdash; This is a new scenario. Unlike installing a package, the models in the `jaffle_finance` project will not be pulled down as source code and parsed into your project. Instead, dbt Cloud provides a metadata service that resolves references to [**public models**](/docs/collaborate/govern/model-access) defined in the `jaffle_finance` project.
```

> 

Reviewer 3: ### Project use case

> 

Reviewer 3: ```suggestion

### Projects use case
```

Resolving the references in this way has several advantages:

> - I'm using an intentional interface, designated by the model's maintainer with `access: public`.

Reviewer 3: ```suggestion

- You're using an intentional interface designated by the model's maintainer with `access: public`.
```

> - I'm keeping the scope of my project narrow. I'm not adding any unneeded resources or complexity. This is faster for me, and faster for dbt.

Reviewer 3: ```suggestion
- You're keeping the scope of your project narrow, and avoiding unnecessary resources and complexity. This is faster for you and faster for dbt.
```

> - I don't need to mirror any conditional configuration of the upstream project: `vars`, environment variables, `target.name`, etc. Wherever the Finance team is building their models in production, that's where I'll be referencing them. The Finance team could re-alias their model, change the name of its schema, or [bump its version](model-versions), and my `ref` would still resolve successfully.

Reviewer 3: ```suggestion
- You don't need to mirror any conditional configuration of the upstream project such as `vars`, environment variables, or `target.name`. You can reference them directly wherever the Finance team is building their models in production. Even if the Finance team makes changes like renaming the model, changing the name of its schema, or [bumping its version](/docs/collaborate/govern/model-versions), your `ref` would still resolve successfully.
```

> - I'm not at risk of accidentally building those models with `dbt run` or `dbt build`. While I can select those models, I cannot actually build them. This avoids the risk of unexpected warehouse costs and permissions issues, and it enables proper ownership and cost allocation for each team's models.

Reviewer 3: ```suggestion
- You eliminate the risk of accidentally building those models with `dbt run` or `dbt build`. While you can select those models, you can't actually build them. This prevents unexpected warehouse costs and permissions issues. This also ensures proper ownership and cost allocation for each team's models.
```

Author: yes!!

Models referenced from a `project`-type dependency must use [two-argument `ref`](/reference/dbt-jinja-functions/ref#two-argument-variant), including the project name:

<File name="models/marts/roi_by_channel.sql">

```sql
with monthly_revenue as (
  
    select * from {{ ref('jaffle_finance', 'monthly_revenue') }}

),

...

```

</File>

> 

Reviewer 3: ### Package use case

> 

Reviewer 3: ```suggestion
### Package use case

```

> If I were to instead install the `jaffle_finance` project as a `package` dependency, this would pull down its full source code. Meaning:

Reviewer 3: ```suggestion
If you were to instead install the `jaffle_finance` project as a `package` dependency, this would pull down its full source code. This means:
```

- dbt needs to parse and resolve more inputs (which is slower)

> - dbt expects me to configure these models as if they were my own (with `vars`, env vars, etc)

Reviewer 3: ```suggestion
- dbt expects you to configure these models as if they were your own (with `vars`, env vars, etc)
```

> - dbt will run these models as my own, unless I explicitly `--exclude` them

Reviewer 3: ```suggestion
- dbt will run these models as your own unless you explicitly `--exclude` them
```

> - I could be using the project's models in a way that their maintainer (the Finance team) hasn't intended

Reviewer 3: ```suggestion
- You could be using the project's models in a way that their maintainer (the Finance team) hasn't intended
```

Installing another internal project as a package can be a useful pattern for:

> - "Unified deployments" in production environments. If the central data platform team of Jaffle Shop wanted to schedule the deployment of models across both `jaffle_finance` and `jaffle_marketing`, using dbt's selection syntax, they could create a new "passthrough" project that installed both projects as packages.

Reviewer 3: ```suggestion
- Unified deployments &mdash; In a production environment, if the central data platform team of Jaffle Shop wanted to schedule the deployment of models across both `jaffle_finance` and `jaffle_marketing`,  they could use dbt's [selection syntax](/reference/node-selection/syntax) to create a new "passthrough" project that installed both projects as packages.
```

> - Making a coordinated change across multiple projects in development. If I wanted to test the effects of a change to a public model in an upstream project (`jaffle_finance.monthly_revenue`), and see how it impacts a downstream model (`jaffle_marketing.roi_by_channel`), before pushing the changes into a staging or production environment, I could install `jaffle_finance` as a package inside `jaffle_marketing`, pointing to a specific git branch. (If I find I regularly need to do end-to-end testing of changes across both projects, I should reexamine if this really represents a stable interface boundary.)

Reviewer 3: ```suggestion
- Coordinated changes &mdash; In development, if you wanted to test the effects of a change to a public model in an upstream project (`jaffle_finance.monthly_revenue`) on a downstream model (`jaffle_marketing.roi_by_channel`) _before_ introducing changes to a staging or production environment, you can install the `jaffle_finance` package as a package within `jaffle_marketing`.  The installation can point to a specific git branch, however, if you find yourself frequently needing to perform end-to-end testing across both projects, we recommend you re-examine if this represents a stable interface boundary. 
```
